app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala

```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
 import play.api.Logger
 import play.api.http.Status._
 import play.api.libs.iteratee.Done
-import play.api.libs.json.Json
+import play.api.libs.json.{JsValue, Json}
 import play.api.libs.streams.Accumulator
 import play.api.mvc.Results._
 import play.api.mvc.{EssentialAction, RequestHeader, Result}
 import uk.gov.hmrc.fileupload.EnvelopeId
-import uk.gov.hmrc.fileupload.transfer.TransferService.{EnvelopeStatusNotFoundError, _}
+import uk.gov.hmrc.fileupload.quarantine.Constraints
+import uk.gov.hmrc.fileupload.transfer.TransferService._
 import uk.gov.hmrc.fileupload.utils.StreamsConverter
 
 import scala.concurrent.{ExecutionContext, Future}
 
 object EnvelopeChecker {
 
-  type WithValidEnvelope = EnvelopeId => EssentialAction => EssentialAction
+  type FileSize = Int
+  type WithValidEnvelope = EnvelopeId => (FileSize => EssentialAction) => EssentialAction
 
   import uk.gov.hmrc.fileupload.utils.StreamImplicits.materializer
 
-  def withValidEnvelope(check: (EnvelopeId) => Future[EnvelopeStatusResult])
+  def withValidEnvelope(check: (EnvelopeId) => Future[EnvelopeDetailResult])
                        (envelopeId: EnvelopeId)
-                       (action: EssentialAction)
+                       (action: FileSize => EssentialAction)
                        (implicit ec: ExecutionContext) =
     EssentialAction { implicit rh =>
       Accumulator.flatten {
         check(envelopeId).map {
-          case Xor.Right("OPEN") =>
-            action(rh)
-          case Xor.Right(otherStatus) =>
-            logAndReturn(LOCKED, s"Unable to upload to envelope: $envelopeId with status: $otherStatus")
-          case Xor.Left(EnvelopeStatusNotFoundError(_)) =>
+          case (Xor.Right(envelope)) =>
+            val status = (envelope \ "status").as[String]
+            if (status == "OPEN") {
+              action(checkSizeLimit(envelope))(rh)
+            } else {
+              logAndReturn(LOCKED, s"Unable to upload to envelope: $envelopeId with status: $status")
+            }
+          case (Xor.Left(EnvelopeDetailNotFoundError(_))) =>
             logAndReturn(NOT_FOUND, s"Unable to upload to nonexistent envelope: $envelopeId")
-          case Xor.Left(error) =>
+          case (Xor.Left(error)) =>
             logAndReturn(INTERNAL_SERVER_ERROR, error.toString)
         }
       }
     }
 
+  def checkSizeLimit(envelope: JsValue) = (
+    envelope \ "constraints").as[Constraints].maxSizePerItem match {
+    case Some(s) =>
+      val fileSize = s.replaceAll("[^\\d.]", "").toInt + 1
+      val fileSizeType = s.toUpperCase.replaceAll("[^KB,MB]{2}", "")
+      fileSizeType match {
+        case "KB" => fileSize * 1024
+        case "MB" => fileSize * 1024 * 1024
+      }
+    case None => 11 * 1024 * 1024
```

Comment:
if we provide a default of 11mb it might be better to be more explicit about it and keep it in a variable. Also when I read this method and its name `check` I assumed return type to be